File: 2023/2.dart
#!/usr/bin/env dart

import 'dart:io';
import 'dart:math';

Map<String, int> limits = {'red': 12, 'green': 13, 'blue': 14};

void part1(List<List<Map<String, int>>> games) {
  int sum = games
      .asMap()
      .map((index, game) => MapEntry(
          index + 1,
          game
              .map((pull) => pull
                  .map((color, value) =>
                      MapEntry(color, (value > limits[color]!) ? false : true))
                  .values
                  .reduce((value, element) => value && element))
              .reduce((value, element) => value && element)))
      .entries
      .where((entry) => entry.value)
      .fold(0, (prev, entry) => prev + entry.key);

  print(sum);
}

void part2(List<List<Map<String, int>>> games) {
  int sum = games
      .map((game) => limits.keys
          .map((color) =>
              game.map((pull) => pull[color] ?? 0).toList().reduce(max))
          .toList()
          .reduce((value, element) => value * element))
      .reduce((value, element) => value + element);

  print(sum);
}

void main() {
  List<String> data = new File("2.input").readAsLinesSync();

  List<List<Map<String, int>>> games = data
      .map((line) => new RegExp(r'Game ([0-9]+): (.*)')
          .firstMatch(line)!
          .group(2)!
          .split(';')
          .map((s) => Map.fromIterable(s.trim().split(', ').map((s) => s.split(' ')),
              key: (item) => item[1].toString(),
              value: (item) => int.parse(item[0])))
          .toList())
      .toList();

  part1(games);

  part2(games);
}

File: 2025/2.py
#!/usr/bin/env python3

def is_invalid_a(value):
    s = str(value)
    n = len(s)

    if n % 2 == 0:
        n //= 2
        return s[:n] == s[n:]
    else:
        return False

def is_invalid_b(value):
    s = str(value)
    n = len(s) // 2

    for i in range(n, 0, -1):
        if len(set(split_str(s, i))) == 1:
            return True

    return False

def split_str(s, n):
    for i in range(0, len(s), n):
            yield s[i:i + n]

def find_invalids(data, valid_func):
    for a, b in data:
        for i in range(a, b + 1):
            if valid_func(i):
                yield i

with open("2.input") as file:
    data = [tuple(map(int, x.split('-'))) for x in file.readline().split(',')]

    print(sum(list(find_invalids(data, is_invalid_a))))
    print(sum(list(find_invalids(data, is_invalid_b))))


